lib/foreman/password_hash.rb

```diff
@@ -0,0 +1,63 @@
+require 'active_support/core_ext/module/delegation'
+
+#
+# Generic-purpose password hasher with two implementations: BCrypt and SHA1.
+#
+# generate_salt(cost) - generates random salt of optional cost (1-30)
+# calculate_salt(object, cost) - calculates hash from given object (useful for tokens)
+# hash_secret(password, salt) - returns hash from secret and salt
+#
+module Foreman
+  class PasswordHash
+    class BCryptImplementation
+      def generate_salt(cost)
+        BCrypt::Engine.generate_salt(cost)
+      end
+
+      def calculate_salt(object, cost)
+        "$2a$#{cost.to_s.rjust(2, '0')}$#{Digest::SHA1.hexdigest(object.to_s)}"
```

Comment:
Would be possible to throw the calculation? Meaning call `generate_salt` in here, to get less dependent on our own implementation for the future?

Reply:
Sorry what you mean? Throw away? For some reason @timogoebel designed the Personal Tokens in a way that salt is a digest of user ID and I wanted to keep this behavior.

Reply:
I believe it does not have to be random. It just has to be unique per user.

Reply:
I believe it is not about being random, but mainly being different, so the password hashes are not similar to hashes of the same password taken elsewhere (or in the same database). Thus I believe the idea is it should be unique per password, not user. But I think being unique per user is safe enough, because we got just one real password (otherwise i think it would be a problem - consider the user is setting same password twice, on both places - the hashes are being same and it gives an attacker first clue about the user's password).